File: src/gate.ts
import { budgetKey, dismissalKey, DAY_SECONDS } from "./checks.js";
import { compilePolicy } from "./policy.js";
import { MemoryStore } from "./stores.js";
import type {
  Candidate,
  Check,
  CheckContext,
  CheckOutcome,
  Decision,
  EvaluateInput,
  GateHooks,
  GateOptions,
  OutcomeEvent,
  Policy,
  Store,
  Surface,
  TraceEntry,
  UserState,
} from "./types.js";

class PrefixedStore implements Store {
  constructor(private readonly inner: Store, private readonly prefix: string) {}
  get(key: string) { return this.inner.get(this.prefix + key); }
  set(key: string, value: string, ttl?: number) { return this.inner.set(this.prefix + key, value, ttl); }
  incr(key: string, ttl?: number) { return this.inner.incr(this.prefix + key, ttl); }
  del(key: string) { return this.inner.del(this.prefix + key); }
}

export interface Gate {
  /** Run every check in order. Never throws for a check failure; see the trace. */
  evaluate(input: EvaluateInput): Promise<Decision>;
  /**
   * Call right before you actually send. Consumes one unit of every budget-like
   * check, in order, and returns false if a unit was taken by a concurrent
   * delivery in the meantime. Idempotent on decision.id: a second call returns
   * the first result without consuming again.
   */
  commit(decision: Decision, input: EvaluateInput): Promise<boolean>;
  /** Tell the gate what happened after delivery, so cooldowns can learn. */
  record(user: Pick<UserState, "id">, candidate: Pick<Candidate, "type">, event: OutcomeEvent, at?: Date): Promise<void>;
  /** Snapshot of the current counters for a user, for debugging and UIs. */
  inspect(user: UserState, now?: Date): Promise<{ budgetUsed: number; dismissals: Record<string, number> }>;
  readonly checks: readonly Check[];
}

/** createGate accepts explicit checks or a JSON policy (see spec/schema/policy.schema.json). */
export interface PolicyGateOptions {
  policy: Policy;
  store?: Store;
  onDecision?: (decision: Decision) => void;
  hooks?: GateHooks;
}

const COMMIT_TTL = 2 * DAY_SECONDS;

export function createGate(options: GateOptions | PolicyGateOptions): Gate {
  if ("policy" in options && "checks" in options) throw new Error("createGate takes either checks or policy, not both");
  const resolved: GateOptions = "policy" in options
    ? { ...compilePolicy(options.policy), ...(options.store ? { store: options.store } : {}), ...(options.onDecision ? { onDecision: options.onDecision } : {}), ...(options.hooks ? { hooks: options.hooks } : {}) }
    : options;
  const store = new PrefixedStore(resolved.store ?? new MemoryStore(), resolved.keyPrefix ?? "pg:");
  const onStoreError = resolved.onStoreError ?? "open";
  const hooks = resolved.hooks ?? {};
  const checks = [...resolved.checks];
  let sequence = 0;
  const consumers = checks.filter((c): c is Check & { consume: NonNullable<Check["consume"]> } => typeof c.consume === "function");

  const callHook = async <K extends keyof GateHooks>(name: K, ctx: CheckContext | null, check: Check | null, ...rest: unknown[]) => {
    const hook = hooks[name] as ((...args: unknown[]) => void | Promise<void>) | undefined;
    if (!hook) return;
    try {
      await hook(...(ctx ? [ctx, check, ...rest] : rest));
    } catch (error) {
      if (name !== "error" && ctx && check) await callHook("error", ctx, check, error);
    }
  };

  const evaluate = async (input: EvaluateInput): Promise<Decision> => {
    const now = input.now ?? new Date();
    const priority = input.candidate.priority ?? "normal";
    const trace: TraceEntry[] = [];
    const shadowed: string[] = [];
    const nearLimit: Decision["nearLimit"] = [];
    let surfaces: Surface[] = pickSurfaces(input.user, input.candidate);
    let deliverAt: Date | undefined;

    const finish = async (partial: Partial<Decision>): Promise<Decision> => {
      const decision: Decision = {
        id: `${input.user.id}:${input.candidate.id}:${now.toISOString()}#${++sequence}`,
        allowed: false,
        userId: input.user.id,
        candidateId: input.candidate.id,
        surfaces: [],
        shadowed,
        nearLimit,
        trace,
        evaluatedAt: now,
        ...partial,
      };
      resolved.onDecision?.(decision);
      await callHook("finally", null, null, decision);
      return decision;
    };

    for (const check of checks) {
      const ctx: CheckContext = { user: input.user, candidate: input.candidate, now, priority, store, surfaces };
      await callHook("before", ctx, check);
      const started = performance.now();
      let outcome: CheckOutcome;
      try {
        outcome = await check.run(ctx);
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        await callHook("error", ctx, check, error);
        if (onStoreError === "closed") {
          trace.push({ id: check.id, outcome: "reject", reason: `check threw (${message}); failing closed`, ms: elapsed(started) });
          return finish({ rejectedBy: check.id, reason: `check "${check.id}" failed and the gate fails closed: ${message}` });
        }
        trace.push({ id: check.id, outcome: "skip", reason: `check threw (${message}); failing open`, ms: elapsed(started) });
        continue;
      }
      const ms = elapsed(started);
      await callHook("after", ctx, check, outcome, ms);
      if (check.nonRejecting && (outcome.kind === "reject" || outcome.kind === "defer")) {
        // A non-rejecting check that tries to stop evaluation is a bug in the check, not a decision about the user.
        trace.push({ id: check.id, outcome: "skip", reason: `non-rejecting check returned ${outcome.kind} (${outcome.reason}); ignored`, ms });
        continue;
      }
      const stops = outcome.kind === "reject" || outcome.kind === "defer";
      const entry: TraceEntry = { id: check.id, outcome: outcome.kind, ms };
      if ("reason" in outcome && outcome.reason) entry.reason = outcome.reason;
      if (stops && check.shadow) entry.shadow = true;
      trace.push(entry);
      if (outcome.kind === "pass" && outcome.nearLimit) nearLimit.push({ check: check.id, ...outcome.nearLimit });
      if (stops && check.shadow) {
        shadowed.push(check.id);
        continue;
      }
      if (outcome.kind === "reject") return finish({ rejectedBy: check.id, reason: outcome.reason });
      if (outcome.kind === "defer") return finish({ deferredBy: check.id, retryAt: outcome.retryAt, reason: outcome.reason });
      if (outcome.kind === "adjust") {
        if (outcome.deliverAt) deliverAt = outcome.deliverAt;
        if (outcome.surfaces) surfaces = outcome.surfaces;
      }
    }
    return finish({ allowed: true, surfaces, ...(deliverAt ? { deliverAt } : {}) });
  };

  const commit = async (decision: Decision, input: EvaluateInput): Promise<boolean> => {
    if (!decision.allowed) return false;
    if (!consumers.length) return true;
    const now = input.now ?? decision.evaluatedAt;
    const priority = input.candidate.priority ?? "normal";
    const marker = `commit:${decision.id}`;
    try {
      const seen = await store.get(marker);
      if (seen !== null) return seen === "1";
      let ok = true;
      for (const check of consumers) {
        const ctx: CheckContext = { user: input.user, candidate: input.candidate, now, priority, store, surfaces: decision.surfaces };
        if (!(await check.consume(ctx))) { ok = false; break; }
      }
      await store.set(marker, ok ? "1" : "0", COMMIT_TTL);
      return ok;
    } catch {
      return onStoreError === "open";
    }
  };

  const record: Gate["record"] = async (user, candidate, event, at = new Date()) => {
    if (event !== "dismissed") return;
    const key = dismissalKey(user.id, candidate.type);
    const raw = await store.get(key);
    const stamps: number[] = raw ? JSON.parse(raw) : [];
    const keepFrom = at.getTime() - 90 * DAY_SECONDS * 1000;
    const next = [...stamps.filter((t) => t >= keepFrom), at.getTime()];
    await store.set(key, JSON.stringify(next), 90 * DAY_SECONDS);
  };

  const inspect: Gate["inspect"] = async (user, now = new Date()) => {
    const budgetUsed = Number((await store.get(budgetKey(user.id, now, user.timezone))) ?? 0);
    const dismissals: Record<string, number> = {};
    for (const type of user.mutedTypes ?? []) {
      const raw = await store.get(dismissalKey(user.id, type));
      dismissals[type] = raw ? (JSON.parse(raw) as number[]).length : 0;
    }
    return { budgetUsed, dismissals };
  };

  return { evaluate, commit, record, inspect, checks };
}

function pickSurfaces(user: UserState, candidate: Candidate): Surface[] {
  const wanted = candidate.surfaces ?? ["feed"];
  if (!user.surfaces) return wanted;
  const allowed = new Set(user.surfaces);
  return wanted.filter((s) => allowed.has(s));
}

const elapsed = (started: number) => Math.round((performance.now() - started) * 1000) / 1000;

File: go/proactivegate/gate.go
package proactivegate

import (
	"encoding/json"
	"fmt"
	"sync/atomic"
	"time"
)

const commitTTLSeconds = 2 * daySeconds

// Gate is a compiled policy: an ordered list of checks bound to a store.
// Create one from explicit checks with NewGate, or from a JSON policy
// document with NewGateFromPolicy.
type Gate struct {
	checks       []Check
	consumers    []Check // the checks with Consume, in policy order
	inner        Store   // the caller's store, unprefixed
	store        Store   // inner under the policy's key prefix
	onStoreError string  // "open" or "closed"
	sequence     atomic.Int64
}

// GateOptions is how a gate is built from checks in code.
type GateOptions struct {
	Checks []Check
	Store  Store
	// OnStoreError decides what a store failure means for a check: "open"
	// lets the candidate through and records the failure in the trace;
	// "closed" rejects. Default "open".
	OnStoreError string
	// KeyPrefix prefixes every key the gate writes. Default "pg:".
	KeyPrefix string
}

// NewGate builds a gate over an explicit check list.
func NewGate(o GateOptions) *Gate {
	store := o.Store
	if store == nil {
		store = NewMemoryStore()
	}
	prefix := o.KeyPrefix
	if prefix == "" {
		prefix = "pg:"
	}
	onStoreError := o.OnStoreError
	if onStoreError == "" {
		onStoreError = "open"
	}
	g := &Gate{
		checks:       append([]Check(nil), o.Checks...),
		inner:        store,
		store:        prefixedStore{inner: store, prefix: prefix},
		onStoreError: onStoreError,
	}
	for _, c := range g.checks {
		if c.Consume != nil {
			g.consumers = append(g.consumers, c)
		}
	}
	return g
}

// Checks returns the compiled check list.
func (g *Gate) Checks() []Check { return g.checks }

// InnerStore returns the caller's store, without the key prefix applied.
func (g *Gate) InnerStore() Store { return g.inner }

func pickSurfaces(user *UserState, candidate *Candidate) []string {
	wanted := candidate.Surfaces
	if len(wanted) == 0 {
		wanted = []string{"feed"}
	}
	if len(user.Surfaces) == 0 {
		return wanted
	}
	var out []string
	for _, s := range wanted {
		if contains(user.Surfaces, s) {
			out = append(out, s)
		}
	}
	return out
}

// Evaluate runs every check in order and never returns an error for a check
// failure; what happened is in the trace.
func (g *Gate) Evaluate(input EvaluateInput) *Decision {
	now := input.Now
	if now.IsZero() {
		now = time.Now()
	}
	priority := input.Candidate.Priority
	if priority == "" {
		priority = PriorityNormal
	}
	var trace []TraceEntry
	var shadowed []string
	var nearLimit []NearLimitEntry
	surfaces := pickSurfaces(input.User, input.Candidate)
	var deliverAt *time.Time

	finish := func(allowed bool, rejectedBy, deferredBy string, retryAt *time.Time, reason string) *Decision {
		d := &Decision{
			ID:          fmt.Sprintf("%s:%s:%s#%d", input.User.ID, input.Candidate.ID, isoMillis(now), g.sequence.Add(1)),
			Allowed:     allowed,
			UserID:      input.User.ID,
			CandidateID: input.Candidate.ID,
			Surfaces:    surfaces,
			Shadowed:    shadowed,
			NearLimit:   nearLimit,
			Trace:       trace,
			EvaluatedAt: now,
			RejectedBy:  rejectedBy,
			DeferredBy:  deferredBy,
			RetryAt:     retryAt,
			Reason:      reason,
		}
		if allowed {
			d.DeliverAt = deliverAt
		} else {
			d.Surfaces = []string{}
		}
		return d
	}

	for _, check := range g.checks {
		started := time.Now()
		ctx := &CheckContext{
			User:      input.User,
			Candidate: input.Candidate,
			Now:       now,
			Priority:  priority,
			Store:     g.store,
			Surfaces:  surfaces,
		}
		outcome, err := check.Run(ctx)
		ms := float64(time.Since(started).Microseconds()) / 1000
		if err != nil {
			if g.onStoreError == "closed" {
				trace = append(trace, TraceEntry{ID: check.ID, Outcome: OutcomeReject, Reason: fmt.Sprintf("check threw (%s); failing closed", err), Ms: ms})
				return finish(false, check.ID, "", nil, fmt.Sprintf("check %q failed and the gate fails closed: %s", check.ID, err))
			}
			trace = append(trace, TraceEntry{ID: check.ID, Outcome: OutcomeSkip, Reason: fmt.Sprintf("check threw (%s); failing open", err), Ms: ms})
			continue
		}
		if check.NonRejecting && (outcome.Kind == OutcomeReject || outcome.Kind == OutcomeDefer) {
			// A non-rejecting check that tries to stop evaluation is a bug in
			// the check, not a decision about the user (3.2).
			trace = append(trace, TraceEntry{ID: check.ID, Outcome: OutcomeSkip, Reason: fmt.Sprintf("non-rejecting check returned %s (%s); ignored", outcome.Kind, outcome.Reason), Ms: ms})
			continue
		}
		stops := outcome.Kind == OutcomeReject || outcome.Kind == OutcomeDefer
		entry := TraceEntry{ID: check.ID, Outcome: outcome.Kind, Reason: outcome.Reason, Ms: ms}
		if stops && check.Shadow {
			entry.Shadow = true
		}
		trace = append(trace, entry)
		if outcome.Kind == OutcomePass && outcome.NearLimit != nil {
			nearLimit = append(nearLimit, NearLimitEntry{Check: check.ID, Used: outcome.NearLimit.Used, Limit: outcome.NearLimit.Limit})
		}
		if stops && check.Shadow {
			shadowed = append(shadowed, check.ID)
			continue
		}
		switch outcome.Kind {
		case OutcomeReject:
			return finish(false, check.ID, "", nil, outcome.Reason)
		case OutcomeDefer:
			return finish(false, "", check.ID, outcome.RetryAt, outcome.Reason)
		case OutcomeAdjust:
			if outcome.DeliverAt != nil {
				at := *outcome.DeliverAt
				deliverAt = &at
			}
			if outcome.Surfaces != nil {
				surfaces = outcome.Surfaces
			}
		}
	}
	return finish(true, "", "", nil, "")
}

// Commit consumes one unit of every budget-like check, in order, and returns
// false when a unit was taken by a concurrent delivery in the meantime (5.2).
// It is idempotent on the decision id: a second call returns the first result
// without incrementing (5.3). Committing a decision that was not allowed
// returns false without touching the store (5.4).
func (g *Gate) Commit(decision *Decision, input EvaluateInput) bool {
	if !decision.Allowed {
		return false
	}
	if len(g.consumers) == 0 {
		return true
	}
	now := input.Now
	if now.IsZero() {
		now = decision.EvaluatedAt
	}
	priority := input.Candidate.Priority
	if priority == "" {
		priority = PriorityNormal
	}
	marker := "commit:" + decision.ID
	seen, ok, err := g.store.Get(marker)
	if err != nil {
		return g.onStoreError == "open"
	}
	if ok {
		return seen == "1"
	}
	commitOK := true
	for _, check := range g.consumers {
		ctx := &CheckContext{
			User:      input.User,
			Candidate: input.Candidate,
			Now:       now,
			Priority:  priority,
			Store:     g.store,
			Surfaces:  decision.Surfaces,
		}
		ok, err := check.Consume(ctx)
		if err != nil {
			return g.onStoreError == "open"
		}
		if !ok {
			commitOK = false
			break
		}
	}
	value := "0"
	if commitOK {
		value = "1"
	}
	if err := g.store.Set(marker, value, commitTTLSeconds); err != nil {
		return g.onStoreError == "open"
	}
	return commitOK
}

// Record tells the gate what happened after delivery so cooldowns can learn.
// Only "dismissed" is recorded; other events are accepted and ignored.
func (g *Gate) Record(userID, candidateType, event string, at time.Time) error {
	if event != "dismissed" {
		return nil
	}
	if at.IsZero() {
		at = time.Now()
	}
	key := dismissalKey(userID, candidateType)
	raw, ok, err := g.store.Get(key)
	if err != nil {
		return err
	}
	var stamps []int64
	if ok {
		if err := json.Unmarshal([]byte(raw), &stamps); err != nil {
			return err
		}
	}
	keepFrom := at.UnixMilli() - 90*daySeconds*1000
	var next []int64
	for _, t := range stamps {
		if t >= keepFrom {
			next = append(next, t)
		}
	}
	next = append(next, at.UnixMilli())
	text, err := json.Marshal(next)
	if err != nil {
		return err
	}
	return g.store.Set(key, string(text), 90*daySeconds)
}
